mne/cov.py

```diff
@@ -1325,65 +1337,72 @@ def prepare_noise_cov(noise_cov, info, ch_names, rank=None,
+    for ch_type, ch_picks in out_idx_type.items():
+        ch_names_type[ch_type] = [ch_names[k] for k in ch_picks]
+        info_picks[ch_type] = []
+        if ch_names_type[ch_type]:
+            # as pick_channels return full list if include is []
+            info_picks[ch_type] = pick_channels(info['ch_names'],
+                                                ch_names_type[ch_type])
+
+    for ch_type, ch_picks in out_idx_type.items():
+        assert (len(info_picks[ch_type]) ==
+                len(ch_names_type[ch_type]) ==
+                len(out_idx_type[ch_type]))
+
+    assert(sum((len(v) for k, v in out_idx_type.items())) == n_chan)
+
     eigvec = np.zeros((n_chan, n_chan))
     eig = np.zeros(n_chan)
-    has_meg = len(out_meg_idx) > 0
-    has_eeg = len(out_eeg_idx) > 0
+    has_type = dict()
+    for ch_type, ch_picks in out_idx_type.items():
+        has_type[ch_type] = len(ch_picks) > 0
 
     # Get the specified noise covariance rank
+    ranks_type = {ch_type: None for ch_type in out_idx_type}
     if rank is not None:
         if isinstance(rank, dict):
-            rank_meg = rank.get('meg', None)
-            rank_eeg = rank.get('eeg', None)
+            for ch_type, ch_picks in out_idx_type.items():
+                ranks_type[ch_type] = rank.get(ch_type, None)
         else:
-            rank_meg = int(rank)
-            rank_eeg = None
-    else:
-        rank_meg, rank_eeg = None, None
-
-    if has_meg:
-        C_meg = C[np.ix_(out_meg_idx, out_meg_idx)]
-        this_info = pick_info(_simplify_info(info), info_pick_meg, copy=False)
-        if rank_meg is None:
-            rank_meg = _estimate_rank_meeg_cov(C_meg, this_info, scalings)
-        eig[out_meg_idx], eigvec[np.ix_(out_meg_idx, out_meg_idx)] = \
-            _get_ch_whitener(C_meg, False, 'MEG', rank_meg)
-    if has_eeg:
-        C_eeg = C[np.ix_(out_eeg_idx, out_eeg_idx)]
-        this_info = pick_info(_simplify_info(info), info_pick_eeg, copy=False)
-        if rank_eeg is None:
-            rank_eeg = _estimate_rank_meeg_cov(C_eeg, this_info, scalings)
-        eig[out_eeg_idx], eigvec[np.ix_(out_eeg_idx, out_eeg_idx)], = \
-            _get_ch_whitener(C_eeg, False, 'EEG', rank_eeg)
-        if _needs_eeg_average_ref_proj(info) and not \
+            ranks_type['meg'] = int(rank)
+
+    for ch_type, this_has in has_type.items():
+        if not this_has:
+            continue
+        this_picks = out_idx_type[ch_type]
+        this_C = C[np.ix_(this_picks, this_picks)]
+        this_info = pick_info(_simplify_info(info), info_picks[ch_type],
+                              copy=False)
+        this_rank = ranks_type[ch_type]
+        if ranks_type[ch_type] is None:
+            this_rank = _estimate_rank_meeg_cov(this_C, this_info, scalings)
+        eig[this_picks], eigvec[np.ix_(this_picks, this_picks)] = \
+            _get_ch_whitener(this_C, False, ch_type.upper(), this_rank)
+
+        # XXX : also handle ref for sEEG and ECoG
```

Comment:
this still needs to be done as you ahve a reference pb too with sEEG or ECoG

Reply:
Let's fix separately #5081